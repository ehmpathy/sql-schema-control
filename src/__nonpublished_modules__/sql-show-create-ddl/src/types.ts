export interface DatabaseConnection {
  query: (args: { sql: string; values?: (string | number)[] }) => Promise<{ rows: any[] }>;
  end: () => Promise<void>;
}
