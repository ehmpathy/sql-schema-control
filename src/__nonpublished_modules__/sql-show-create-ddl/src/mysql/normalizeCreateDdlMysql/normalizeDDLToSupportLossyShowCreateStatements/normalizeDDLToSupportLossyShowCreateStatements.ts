import sqlFormatter from 'sql-formatter';
import strip from 'sql-strip-comments';

import { ResourceType } from '../../../../../../types';
import { recursivelyHeavilyNormalizeViewDdl } from '../../../generic/recursivelyHeavilyNormalizeViewDdl/recursivelyHeavilyNormalizeViewDdl';

const RESOURCES_WITH_LOSSY_SHOW_CREATE_STATEMENTS = [ResourceType.TABLE, ResourceType.VIEW];

/**
 * some SHOW CREATE statements are not lossy:
 * - procedures and functions preserve formatting and comments, making getting the sql diff a dream
 *
 * some SHOW CREATE statements are lossy:
 * - tables strip comments and apply their own standard formatting
 * - views strip comments and remove all newlines (which technically _is_ a special kind of formatting) - and apply all sorts of fun additions
 */
export const normalizeDDLToSupportLossyShowCreateStatements = ({
  ddl,
  resourceType,
}: {
  ddl: string;
  resourceType: ResourceType;
}) => {
  if (!RESOURCES_WITH_LOSSY_SHOW_CREATE_STATEMENTS.includes(resourceType)) return ddl; // if does not have lossy show create, return the ddl without formatting

  // strip comments from tables and views, since these are not preserved by the SHOW CREATEs
  const strippedDDL = strip(ddl);

  // if view, fix a few fun "features" that the SHOW CREATE statement adds
  const extraNormalizedDDL =
    resourceType === ResourceType.VIEW
      ? recursivelyHeavilyNormalizeViewDdl({ ddl: strippedDDL }) // SHOW CREATE for view has a lot of fun features, so extra normalization is needed
      : strippedDDL;

  // apply formatting, since views have bad formatting
  const formattedDDL = sqlFormatter.format(extraNormalizedDDL);

  // return the cleaned ddl
  return formattedDDL;
};
