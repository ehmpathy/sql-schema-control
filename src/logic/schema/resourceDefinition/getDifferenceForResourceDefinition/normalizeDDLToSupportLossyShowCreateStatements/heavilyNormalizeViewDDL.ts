import {
  flattenSqlByReferencingAndTokenizingSubqueries,
  hydrateSqlByReferencingAndReplacingSubqueryTokens,
} from '../../../../../__nonpublished_modules__/sql-subquery-tokenizer';

// TODO: find a better way of doing this rather than all this ad hock normalization... or atleast refactor this normalization to make it less dependent on previous transformation (maybe split parts into their own fns)
// TODO: support subqueries. right now we are not normalizing them. ideally we would flatten and run normalization on root + all subqueries, and then merge back
export const heavilyNormalizeViewDDL = ({ ddl }: { ddl: string }) => {
  // 0. make all "as" words, except the first, lowercase (for some reason, SHOW CREATE lowercases all tokens EXCEPT `as`, which it explicitly upper cases...)
  const casedAsDdl = ddl
    .replace(/ AS /g, ' as ') // /g to apply to each
    .replace(/ as /, ' AS '); // no /g so only apply to first case

  // 1. lowercase everything in the query, since show create does :upside_down_smiley:
  const casedAsDdlParts = casedAsDdl.split(' AS ');
  const casedDdl = [casedAsDdlParts[0], casedAsDdlParts[1].toLowerCase()].join(' AS ');

  // 2. strip out all of the back ticks. until we have an example where that breaks something, they're just too much to force users to have to maintain
  const ddlWithoutBackticks = casedDdl.replace(/`/g, '');

  // 3. strip out all of the parenthesis (since SHOW CREATE adds them everywhere and thats too much to force users to maintain); unideal since it modifies the meaning, but this is only relevant for diffing
  const { flattenedSql: flattenedDdl, references } = flattenSqlByReferencingAndTokenizingSubqueries({
    sql: ddlWithoutBackticks,
  });
  const whereCasing = !!flattenedDdl.match(/\sWHERE\s/) ? 'WHERE' : 'where'; // determine if user is using upper case or lower case "where"
  const ddlSplitOnWhere = ddlWithoutBackticks.split(whereCasing);
  if (ddlSplitOnWhere.length > 2) {
    throw new Error('found more than two parts of DDL after splitting on where; unexpected'); // fail fast
  }
  const ddlWithoutParens =
    ddlSplitOnWhere.length === 2
      ? ddlSplitOnWhere[0] + '\n' + whereCasing + '\n' + ddlSplitOnWhere[1].replace(/[\(\)]/g, '') // tslint:disable-line prefer-template
      : ddlSplitOnWhere[0]; // if no where clause, then nothing to replace
  const rehydratedDdl = hydrateSqlByReferencingAndReplacingSubqueryTokens({
    flattenedSql: ddlWithoutParens,
    references,
  });
  const ddlWithoutParensInWhereConditions = rehydratedDdl;

  // 4. strip out the "double parenthesis" that SHOW CREATE likes to put on the "join on" statements
  const ddlWithoutDoubleParens = ddlWithoutParensInWhereConditions.replace(/\(\(/g, ' ').replace(/\)\)/g, ' ');

  // 5. strip out the first and last paren of the "FROM" section, IF it is defined; show create likes to wrap the whole "FROM" section w/ parens...
  const ddlHasParenOpenRightAfterFromClause = !!ddlWithoutDoubleParens.match(/\sfrom\s+\(/gi);
  let ddlWithoutParenStartingAndEndingFromClause = ddlWithoutDoubleParens;
  if (ddlHasParenOpenRightAfterFromClause) {
    // flatten the ddl so that we only have one FROM clause
    const { flattenedSql: flattenedDdl, references } = flattenSqlByReferencingAndTokenizingSubqueries({
      sql: ddlWithoutDoubleParens,
    });

    // split DDL on the from statement, since we know it exists
    const flattenedDdlParts = flattenedDdl.split(/\sfrom\s/gi);
    if (flattenedDdlParts.length !== 2) throw new Error('not exactly two parts after splitting on from; unexpected'); // fail fast

    // remove the parens from beginning and end
    const fromClauseWithoutFirstOrLastParens = flattenedDdlParts[1].replace(/^\s*\(/, ' ').replace(/\)\s*$/, ' ');

    // join them back and the ddl w/o this mess
    const flattenedDdlWithoutOpenCloseParenInFromClause =
      flattenedDdlParts[0] + '\nfrom\n' + fromClauseWithoutFirstOrLastParens; // tslint:disable-line prefer-template

    // rehydrate them
    const hydratedDdlWithoutOpenCloseParenInFromClause = hydrateSqlByReferencingAndReplacingSubqueryTokens({
      flattenedSql: flattenedDdlWithoutOpenCloseParenInFromClause,
      references,
    });

    // and set the variable accessible to full fn scope
    ddlWithoutParenStartingAndEndingFromClause = hydratedDdlWithoutOpenCloseParenInFromClause;
  }

  // 6. strip out the final `;` if it exists
  const ddlWithoutFinalSemicolon = ddlWithoutParenStartingAndEndingFromClause.replace(/;/g, '');

  // 7. replace `,(` patterns w/ space in between, since our formatter downstream does not like that
  const ddlWithoutTouchingCommaParen = ddlWithoutFinalSemicolon.replace(/,\(/g, ', (');

  // 8. return that result
  return ddlWithoutTouchingCommaParen;
};
