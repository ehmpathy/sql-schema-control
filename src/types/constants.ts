export enum DefinitionType {
  RESOURCE = 'resource',
  CHANGE = 'change',
}

export enum ResourceType {
  TABLE = 'TABLE',
  FUNCTION = 'FUNCTION',
  PROCEDURE = 'PROCEDURE',
  VIEW = 'VIEW',
}

export enum DatabaseLanguage {
  MYSQL = 'mysql',
}

export interface DatabaseConnection {
  query: (args: { sql: string; values?: any[] }) => Promise<any>;
  end: () => Promise<void>;
}

export enum ChangeDefinitionStatus {
  NOT_APPLIED = 'NOT_APPLIED',
  UP_TO_DATE = 'UP_TO_DATE',
  OUT_OF_DATE = 'OUT_OF_DATE',
}
export enum ResourceDefinitionStatus {
  NOT_APPLIED = 'NOT_APPLIED',
  SYNCED = 'SYNCED',
  OUT_OF_SYNC = 'OUT_OF_SYNC',
  NOT_CONTROLLED = 'NOT_CONTROLLED',
}

export enum RequiredAction { // i.e., the the action we need to take
  NO_CHANGE = 'NO_CHANGE',
  APPLY = 'APPLY',
  REAPPLY = 'REAPPLY', // e.g., display yellow because we can fix (e.g., reapply) the discrepancy
  MANUAL_REAPPLY = 'MANUAL_REAPPLY', // e.g., display red because we cant automatically fix the discrepancy
  MANUAL_MIGRATION = 'MANUAL_MIGRATION', // e.g., display red because we cant automatically fix the discrepancy
  MANUAL_PULL = 'MANUAL_PULL', // e.g., we've found an uncontrolled resource and it should be pulled or deleted
}
