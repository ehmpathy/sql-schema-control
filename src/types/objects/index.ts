export { ChangeDefinition } from './ChangeDefinition';
export { ConnectionConfig } from './ConnectionConfig';
export { ControlConfig } from './ControlConfig';
export { ControlContext } from './ControlContext';
