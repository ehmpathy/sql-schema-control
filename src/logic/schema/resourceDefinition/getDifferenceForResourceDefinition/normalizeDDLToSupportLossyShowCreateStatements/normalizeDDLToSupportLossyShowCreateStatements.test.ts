import { normalizeDDLToSupportLossyShowCreateStatements } from './normalizeDDLToSupportLossyShowCreateStatements';
import { ResourceType } from '../../../../../types';

describe('normalizeDDLToSupportLossyShowCreateStatements', () => {
  it('should strip comments and reformat table create ddl', () => {
    const cleaned = normalizeDDLToSupportLossyShowCreateStatements({
      ddl: `
        CREATE TABLE \`some_resource_table\` (
          \`id\` int(11) NOT NULL AUTO_INCREMENT, -- this id is the primary key, which we define below
          \`created_at\` datetime DEFAULT CURRENT_TIMESTAMP,
          PRIMARY KEY (\`id\`)
          ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci
          `.trim(),
      resourceType: ResourceType.TABLE,
    });
    expect(cleaned).not.toContain('this id is the primary key');
    expect(cleaned).toMatchSnapshot(); // and check snapshot for formatting
  });
  it('should strip comments and reformat view create ddl', () => {
    const cleaned = normalizeDDLToSupportLossyShowCreateStatements({
      ddl: `
CREATE VIEW view_something_for_getResourceCreateStatement AS
SELECT
  'hello, world!' as first_words -- because we want to keep this
  ;
          `.trim(),
      resourceType: ResourceType.TABLE,
    });
    expect(cleaned).not.toContain('because we want to keep this');
    expect(cleaned).toMatchSnapshot(); // and check snapshot for formatting
  });
  it('should normalize view create ddl so that custom user def and show create def match', () => {
    const cleanedUserDef = normalizeDDLToSupportLossyShowCreateStatements({
      ddl: `
CREATE VIEW view_something_for_getResourceCreateStatement AS
select -- test comments
  'hello, world!' as first_words,
  u.id as id
from test_table_getResourceCreateStatement u
where 1=1
    and u.id = 5
          `.trim(),
      resourceType: ResourceType.VIEW,
    });
    const cleanedShowCreateDef = normalizeDDLToSupportLossyShowCreateStatements({
      ddl:
        "CREATE VIEW `view_something_for_getResourceCreateStatement` AS select 'hello, world!' AS `first_words`,`u`.`id` AS `id` from `test_table_getResourceCreateStatement` `u` where ((1 = 1) and (`u`.`id` = 5))",
      resourceType: ResourceType.VIEW,
    });
    expect(cleanedUserDef).toEqual(cleanedShowCreateDef);
  });
});
