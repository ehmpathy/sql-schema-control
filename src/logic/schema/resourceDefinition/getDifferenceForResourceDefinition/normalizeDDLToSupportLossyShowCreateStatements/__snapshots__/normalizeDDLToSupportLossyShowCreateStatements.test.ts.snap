// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`normalizeDDLToSupportLossyShowCreateStatements should strip comments and reformat table create ddl 1`] = `
"CREATE TABLE \`some_resource_table\` (
  \`id\` int(11) NOT NULL AUTO_INCREMENT,
  \`created_at\` datetime DEFAULT CURRENT_TIMESTAMP,
  PRIMARY KEY (\`id\`)
) ENGINE = InnoDB DEFAULT CHARSET = utf8mb4 COLLATE = utf8mb4_unicode_ci"
`;

exports[`normalizeDDLToSupportLossyShowCreateStatements should strip comments and reformat view create ddl 1`] = `
"CREATE VIEW view_something_for_getResourceCreateStatement AS
SELECT
  'hello, world!' as first_words;"
`;
